File: Maximum Subarray.cpp
//dp 

#define F(i, a, b) for(int i = a; i <= b; i++)
#include <vector>
#include <unordered_set>

class Solution {
public:
    int maxSubArray(int A[], int n) {
        int max = INT_MIN;
        int sum = 0;

        F(i, 0, n-1){
            sum += A[i];
            max = max>sum?max:sum;
            sum = sum>0?sum:0;
        }

        return max;
    }
};

File: LeetCode-Java/Maximum Subarray.java
public class Solution {
	public static int maxSubArray(int[] Ary) {
		int len = Ary.length;

		if( len < 1 ){
			return 0;
		}

		//int [] dp = new int [len];
		int max = Integer.MIN_VALUE;
		int sum = 0;
		for( int i = 0; i < len; i++ ){
			sum += Ary[i];
			max = Math.max(sum, max);
			sum = Math.max(sum, 0);
		}

		return max; 
	}
}
